### PR TITLE
Move binarySearch to foundation.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

This gets us closer to fewer dependencies, and removes an indirection used in package:collection.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/animated_list.dart [hunk 1]
```diff
@@ -228,15 +226,34 @@ class AnimatedListState extends State<AnimatedList> with TickerProviderStateMixi
...
     return i == -1 ? null : items.removeAt(i);
   }
 
   _ActiveItem _activeItemAt(List<_ActiveItem> items, int itemIndex) {
-    final int i = binarySearch(items, _ActiveItem.index(itemIndex));
+    final int i = _binarySearch(items, _ActiveItem.index(itemIndex));
     return i == -1 ? null : items[i];
   }
 
+  static int _binarySearch<T extends Comparable<T>>(List<T> sortedList, T value) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/animated_list.dart]: I wonder if this should live in foundation/collections.dart?

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/animated_list.dart]: I don't think we should add to the foundation API.

**Comment by jonahwilliams** [comment]
I ran a micro-benchmark and found binary search to be faster than indexWhere. I also updated the type constraint to be more flexible. Because `int implements Comparable<num>`, the previous constrain disallows numbers

**Comment by jonahwilliams** [comment]
cc @goderbauer

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**